src/main/java/org/simplify4u/plugins/ValidationChecksum.java

```diff
@@ -0,0 +1,136 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.simplify4u.plugins;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.plugin.logging.Log;
+import org.bouncycastle.crypto.digests.SHA256Digest;
+import org.bouncycastle.pqc.math.linearalgebra.ByteUtils;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.Arrays;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * ValidationChecksum is a checksum of a deterministic collection of artifacts.
+ *
+ * The checksum can be used to check against a stored value of a prior validation and for
+ * itself to be stored once full validation has completed.
+ */
+final class ValidationChecksum {
+
+    private final Log log;
+    private final File file;
+    private final byte[] checksum;
+
+    private ValidationChecksum(Log log, File file, byte[] checksum) {
+        this.log = requireNonNull(log);
+        this.file = requireNonNull(file);
+        this.checksum = requireNonNull(checksum);
+    }
+
+    /**
+     * Compare current checksum against previously stored checksum value.
+     *
+     * @return Returns true iff checksum for previous run exists and is equal.
+     */
+    boolean checkValidation() {
+        final byte[] checksumPriorValidation;
+        try {
+            log.debug("Looking for prior validation checksum at: " + file.getAbsolutePath());
+            checksumPriorValidation = FileUtils.readFileToByteArray(file);
+        } catch (final FileNotFoundException e) {
+            log.debug("No prior validation executed. Continuing with full artifact validation.");
+            return false;
+        } catch (final IOException e) {
+            log.debug("Validation of artifacts against prior validation run failed.", e);
+            return false;
+        }
+
+        return Arrays.equals(this.checksum, checksumPriorValidation);
+    }
+
+    /**
+     * Save current checksum to file.
+     */
+    void saveChecksum() {
+        try {
+            FileUtils.writeByteArrayToFile(file, checksum);
+        } catch (final IOException e) {
+            log.debug("Failed to save checksum after successful artifact validation.", e);
+        }
+    }
+
+    /**
+     * A builder for the ValidationChecksum.
+     */
+    static final class Builder {
+
+        private static final String FILENAME_CHECKSUM_PRIOR_VALIDATION = "pgpverify-prior-validation-checksum";
+
+        private final Log log;
+        private final File file;
+
+        /**
+         * Constructor for a builder that uses the given logger and performs file system
+         * operations in a given directory.
+         *
+         * @param log the logger instance
+         * @param directory the target directory
+         */
+        Builder(Log log, File directory) {
```

Comment:
This builder approach make code a little difficult to read ... 

maybe simply in checkValidation method call two method one for read file and one for calculate checksum

Reply:
That's how it started out, but that means we do the checksum calculation twice. Once when validating against previous checksum, and once when producing checksum to be saved.

If we keep checksumming internal to the class, we can easily change it, so I prefer to keep it as a builder. The builder basically managing: 1. the checksum filename, and 2. the (pre)calculated checksum.

Again, let me know whether you agree.

Reply:
agree